Automattic/sensei
# Prevent double form submission

### Overview

This fix tries to reduce the occurance of issue #2748.

The issue happens because we can't guarantee that there will be a single `sensei_course_status` and `sensei_lesson_status` for each user and course/lesson in `wp_comments` table. The checks done [here](https://github.com/Automattic/sensei/blob/8f7cb5b973c8de53a6c3a4ce5441544c9ba5c842/includes/class-sensei-utils.php#L74) do not work if they are executed in parallel. A proper fix would require to move these data in a custom table which would allow us to define proper constraints. Until a proper fix is made, this PR aims to reduce the amount of duplicates that are created.

Most of the issues reported are about duplicates for `sensei_lesson_status`. After code inspection it seems that these statuses are created after user interaction so the only way that I can think of having parallel requests for the same user is by double form submission (e.g. if the user double clicks when cliking a button). Firefox prevents a form being submitted twice on a double click. 

This PR adds a check in the buttons which eventually create a `sensei_lesson_status` to prevent more than one submissions. Although there is still a chance to have double form submissions, I expect that after this change they will be less frequent.

### Changes proposed in this Pull Request

* Adds a script which allows only one form submission every 2 seconds.
* Enqueues the script in lesson, quiz and course pages in the frontend and in learner management in admin.

### Testing instructions

* It wasn't possible to reproduce the actual issue so testing should ensure that it is not possible to have multiple submissions by clicking the buttons many times. I used logging in the PHP side to test it.
* The buttons affected are:
    * The three buttons (Reset, Save, Complete) in quiz page.
    * The complete lesson button in lesson page.
    * The 'Mark as complete' button in 'My courses' and course page.
    * The 'Apply' button in the popup in bulk actions page in admin.
